### PR TITLE
HIVE-28623: Implement RuntimeContext to decouple runtime data from TezTask

### PR DESCRIPTION
**Pull request by abstractdog**
### What changes were proposed in this pull request?
Encapsulate runtime related info to a separate class for later usage.

### Why are the changes needed?
Code refactor.

### Does this PR introduce _any_ user-facing change?
No.

### Is the change a dependency upgrade?
No.

### How was this patch tested?
Easy refactor, precommit tests should pass.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java [hunk 1]
```diff
@@ -125,6 +130,12 @@ public QueryPlan getPlan() {
 
   public void setPlan(QueryPlan plan) {
     this.plan = plan;
+    TezTask task = Utilities.getFirstTezTask(plan.getRootTasks()).orElse(null);
```
Comment (tanishq-chugh) [inline ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:133]: @abstractdog  After query run completes, when resources are being released by the driver, plan is being set to null and it is causing NullPointerException here, which is present as part of DEBUG logs.

Comment (abstractdog) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:133]: I see, can you point me to the logs that show this?

Comment (tanishq-chugh) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:133]: `0: jdbc:hive2://localhost:10000> show tables;
INFO : Compiling command(queryId=tanishqchugh_20241113082511_604beac2-0686-4de3-93e2-edb71629232c): show tables
INFO : Semantic Analysis Completed (retrial = false)
INFO : Created Hive schema: Schema(fieldSchemas:[FieldSchema(name:tab_name, type:string, comment:from deserializer)], properties:null)
INFO : Completed compiling command(queryId=tanishqchugh_20241113082511_604beac2-0686-4de3-93e2-edb71629232c); Time taken: 0.071 seconds
INFO : Executing command(queryId=tanishqchugh_20241113082511_604beac2-0686-4de3-93e2-edb71629232c): show tables
INFO : PREHOOK: query: show tables
INFO : PREHOOK: type: SHOWTABLES
INFO : PREHOOK: Input: database:default
INFO : Starting task [Stage-0:DDL] in serial mode
DEBUG : Task getting executed using mapred tag : tanishqchugh_20241113082511_604beac2-0686-4de3-93e2-edb71629232c,userid=tanishqchugh
INFO : POSTHOOK: query: show tables
INFO : POSTHOOK: type: SHOWTABLES
INFO : POSTHOOK: Input: database:default
INFO : Completed executing command(queryId=tanishqchugh_20241113082511_604beac2-0686-4de3-93e2-edb71629232c); Time taken: 0.105 seconds
DEBUG : Exception while clearing the Fetch task
java.lang.NullPointerException: null
	at org.apache.hadoop.hive.ql.DriverContext.setPlan(DriverContext.java:133) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.Driver.releasePlan(Driver.java:794) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.Driver.releaseResources(Driver.java:611) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:214) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:143) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:138) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hadoop.hive.ql.reexec.ReExecDriver.run(ReExecDriver.java:190) ~[hive-exec-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hive.service.cli.operation.SQLOperation.runQuery(SQLOperation.java:234) ~[hive-service-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hive.service.cli.operation.SQLOperation.access$500(SQLOperation.java:88) ~[hive-service-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at org.apache.hive.service.cli.operation.SQLOperation$BackgroundWork$1.run(SQLOperation.java:334) ~[hive-service-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_422]
	at javax.security.auth.Subject.doAs(Subject.java:422) ~[?:1.8.0_422]
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1899) ~[hadoop-common-3.3.6.jar:?]
	at org.apache.hive.service.cli.operation.SQLOperation$BackgroundWork.run(SQLOperation.java:354) ~[hive-service-4.1.0-SNAPSHOT.jar:4.1.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_422]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_422]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_422]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_422]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_422]
DEBUG : Shutting down query show tables
+------------------+
|   tab_name   |
+------------------+
| customer     |
| employees    |
| final_employees |
| lineitem     |
| nation      |
| orders      |
| part       |
| partsupp     |
| region      |
| supplier     |
| temp       |
+------------------+
11 rows selected (0.258 seconds)
0: jdbc:hive2://localhost:10000>`

Comment (tanishq-chugh) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:133]: Sure
[temp.txt](https://github.com/user-attachments/files/17745294/temp.txt)

Comment (tanishq-chugh) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:133]: this shouldn't cause any issue wrt query execution but just some noise in logs. a null check wrt plan should just be fine.. i guess the one test failure in precommit is as well linked to this ;)

File: ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java [hunk 2]
```diff
@@ -125,6 +130,17 @@ public QueryPlan getPlan() {
 
   public void setPlan(QueryPlan plan) {
     this.plan = plan;
+    // only set runtimeContext if the plan is not null
+    // we don't want to nullify runtimeContext if this method is called with plan=null, which is the case when e.g.
+    // driver.releasePlan() tries to release resources/objects that are known to be heavy
+    if (plan != null){
```
Comment (difin) [inline ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:136]: nit: missing space between `){`

**Comment by difin** [comment]
LGTM +1

File: ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java (hunk 2 above)
Comment (abstractdog) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:136]: ack, let me reformat

**Comment by abstractdog** [comment]
> LGTM +1

thanks for the review, pushed one more commit to address that

File: ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java [hunk 3]
```diff
@@ -125,6 +130,17 @@ public QueryPlan getPlan() {
 
   public void setPlan(QueryPlan plan) {
     this.plan = plan;
+    // only set runtimeContext if the plan is not null
+    // we don't want to nullify runtimeContext if this method is called with plan=null, which is the case when e.g.
+    // driver.releasePlan() tries to release resources/objects that are known to be heavy
+    if (plan != null) {
+      TezTask task = Utilities.getFirstTezTask(plan.getRootTasks()).orElse(null);
+      this.runtimeContext = task == null ? null : task.getRuntimeContext();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java:138]: minor style: i usually wrap ternary condition in round brackets, unless it's single boolean var 
````
this.runtimeContext = (task == null) ? null : task.getRuntimeContext();
````

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java [hunk 4]
```diff
@@ -674,6 +687,13 @@ DAGClient submit(DAG dag, Ref<TezSessionState> sessionStateRef) throws Exception
     return new SyncDagClient(dagClient);
   }
 
+  private DAGClient submitInternal(DAG dag, TezSessionState sessionState) throws TezException, IOException {
+    TezClient tezClient = sessionState.getSession();
+    DAGClient dagClient = tezClient.submitDAG(dag);
+    runtimeContext.initFromTezClient(tezClient);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:693]: shouldn't the runtimeContext be part of TezSessionState?

Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:693]: wouldn't it be better to init
````
runtimeContext.initFromTezClient(sessionState.getSession())
````
before we return the SyncDagClient?

Comment (abstractdog) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:693]: ack, that way we don't need the extra method

**Comment by abstractdog** [comment]
thanks @deniskuzZ, addressed comments in https://github.com/apache/hive/pull/5538/commits/cd391822537e7a6dfeed5b7944990823c53ef5b6

**Review by deniskuzZ — APPROVED** [review]
+1

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java [hunk 5]
```diff
@@ -671,6 +684,7 @@ DAGClient submit(DAG dag, Ref<TezSessionState> sessionStateRef) throws Exception
     }
 
     perfLogger.perfLogEnd(CLASS_NAME, PerfLogger.TEZ_SUBMIT_DAG);
+    runtimeContext.initFromTezClient(sessionState.getSession());
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:687]: minor: why not simply `runtimeContext.init(sessionState.getSession())`

Comment (abstractdog) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezTask.java:687]: agreed, let me rewrite (I think in the early implementation phase I thought there will be different init methods), init looks cleaner

**Pull request merged**